text/2022-07-25-PROPOSAL-build-glyphs-in-rust.md

```diff
@@ -0,0 +1,128 @@
+
+Complicating this, the core approach we take to build a variable font is excitingly indirect (https://simoncozens.github.io/compiling-variable-fonts/):
+
+   1. Generate static TTFs at master locations
+   1. Merge those TTFs to create a final VF
+
+As far as the author can tell there is no requirement for the statics to ever exist, we do it this way purely because
+it was a relatively simple way to get fontmake to support VF. In the end we should go from sources → final VF without this step.
+
+A concrete example may help, consider the N in [Josefin Slab](https://fonts.google.com/specimen/Josefin+Slab)
+(https://github.com/TypeNetwork/Josefinslab) upright:
+
+* [sources/JosefinSlab-Thin.ufo/glyphs/N_.glif](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab-Thin.ufo/glyphs/N_.glif) defines the outline contours at thin (N)
+* [sources/JosefinSlab-Bold.ufo/glyphs/N_.glif](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab-Bold.ufo/glyphs/N_.glif) defines the outline contours at bold (N)
+* [sources/JosefinSlab.designspace](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab.designspace) defines how the parts fit together
+   * JosefinSlab-Thin.ufo is at weight 14
+   * JosefinSlab-Bold.ufo is at weight 83
+   * Font weights are mapped onto the space defined by the UFO
+      * Default 100 maps to 14
+      * 300 maps to 28, 400 to 41, etc
+         * These aren't where you'd land by lerp, we need to watch how we set up interpolation
+
+### How do you do it faster
+
+#25 suggests that build time scales significantly with two things:
+
+1. The amount of per glyph work, #glyphs * #masters, the primary predictor of build time
+   * Makes sense, we do that work glyph by glyph ([varLib._add_gvar](https://github.com/fonttools/fonttools/blob/455158f2bfd9eae2135a5f85bb96549a545cac82/Lib/fontTools/varLib/__init__.py#L222)) and the per glyph work scales directly based on how many masters you have to worry about
+1. The complexity of layout, which for complex scripts can be a very significant factor
+
+Paraphrasing things Simon Cozens’ has claimed for some time, you do three things:
+
+1. Do the work faster, such as by doing it in Rust instead of Python
+1. Do the work concurrently wherever possible
+1. Compute the final result as directly as possible
+   * Don't do things like build a binary font for each master
+
+Since we convert glyphs files to UFO lets focus on how a UFO should build (the actual code should abstract some of the concepts):
+
+1. Concurrently process each glyph, plus the feature file(s)
+   1. For Josefin Slab, process all N_.glif files together to produce the glyph and varstore you need for just N
+      1. `f(all N_.gif, JosefinSlab.designspace) → glyf,gvar for the N`
+   1. Merge all the feature files and compile them
+      1. Shouldn't need final outlines to compute features
+      1. `f(groups.plist, kerning.plist, features.fea, glyphs/contents.plist) → layout table(s)`
```

Comment:
Also needs anchors from all glyphs, though geometry is not strictly necessary iirc.

Reply:
> Shouldn't need final outlines to compute features

currently no, but there's a proposal to support vmtx/VORG overrides in features.fea in for fontTools.feaLib that would require access to outlines in order to retrieve (glyf) or compute (CFF) the bbox.yMax needed to calculate top sidebearings and vertical origin.
https://github.com/fonttools/fonttools/pull/2704

I would argue that is niche enough that can be ignored for step one.

Reply:
Good point. features is step 2+ so I think that doesn't impede step 1. If features has to come after a parallel glyph build that's tiresome but ok.

Reply:
I'll refine the comment re outlines

Reply:
Updated md w/feedback, much appreciated.